File: node/src/csharp_runner.js
// CSharpNuGetScanner.js
import fs   from "fs";
import path from "path";

/**
 * Scans C# / .NET projects for installed NuGet packages.
 *
 * Detection strategy:
 *   1.  A directory is a NuGet project root when it contains at least one
 *       *.csproj / *.fsproj / *.vbproj file.
 *   2.  Installed packages are read from:
 *         • packages.lock.json   (NuGet lock file, preferred)
 *         • obj/project.assets.json  (MSBuild restore fallback)
 *   3.  Central Package Management: Directory.Packages.props is walked up
 *       from the project root to resolve versions missing from the .csproj.
 *   4.  Multi-target projects: the highest resolved version across TFMs is used.
 *   5.  PURL: pkg:nuget/<name>@<version>
 *            (no @ when version is unknown)
 */
export class CSharpNuGetScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // Omits @<version> when version is blank to avoid silent PURL collisions.
  // ─────────────────────────────
  static _nugetPurl(name, version) {
    const base = `pkg:nuget/${name.toLowerCase()}`;
    return version ? `${base}@${version}` : base;
  }

  // ─────────────────────────────
  // Detect .NET project root
  // ─────────────────────────────
  static _isDotnetRoot(dir) {
    try {
      return fs.readdirSync(dir).some(f => /\.(cs|fs|vb)proj$/.test(f));
    } catch {
      return false;
    }
  }

  // ─────────────────────────────
  // Walk ancestors for Directory.Packages.props (Central Package Management).
  // Returns Map<lowercaseName, version> or empty Map.
  // ─────────────────────────────
  static _readCentralPackageProps(startDir) {
    const versions = new Map();
    let dir = startDir;

    for (let i = 0; i < 8; i++) {                 // cap ancestor search depth
      const candidate = path.join(dir, "Directory.Packages.props");
      if (fs.existsSync(candidate)) {
        let xml;
        try { xml = fs.readFileSync(candidate, "utf8"); } catch { break; }
        // <PackageVersion Include="Foo.Bar" Version="1.2.3" />
        const re = /<PackageVersion[\s\S]*?Include=["']([^"']+)["'][\s\S]*?Version=["']([^"']+)["']/gi;
        let m;
        while ((m = re.exec(xml)) !== null) {
          versions.set(m[1].toLowerCase(), m[2]);
        }
        break;                                    // stop at first props file found
      }
      const parent = path.dirname(dir);
      if (parent === dir) break;                  // filesystem root
      dir = parent;
    }

    return versions;
  }

  // ─────────────────────────────
  // Parse packages.lock.json (preferred).
  // Merges all TFMs, keeping the highest version when the same package appears
  // under multiple target frameworks.
  // Returns Map<lowercaseName, { name, version, type, dependencies[] }> or null.
  // ─────────────────────────────
  static _readPackagesLock(projectRoot) {
    const lockPath = path.join(projectRoot, "packages.lock.json");
    if (!fs.existsSync(lockPath)) return null;

    let raw;
    try {
      raw = JSON.parse(fs.readFileSync(lockPath, "utf8"));
    } catch {
      return null;
    }

    const index = new Map();

    for (const tfmDeps of Object.values(raw.dependencies ?? {})) {
      for (const [pkgId, meta] of Object.entries(tfmDeps)) {
        const key     = pkgId.toLowerCase();
        const version = meta.resolved ?? meta.requested ?? "";
        if (!version) continue;                   // skip malformed entries

        if (!index.has(key)) {
          index.set(key, {
            name:         pkgId,
            version,
            type:         (meta.type ?? "direct").toLowerCase(),
            dependencies: Object.keys(meta.dependencies ?? {}).map(d => d.toLowerCase())
          });
        } else {
          // Keep the highest version seen across TFMs
          const existing = index.get(key);
          if (this._versionGt(version, existing.version)) {
            existing.version = version;
          }
        }
      }
    }

    return index.size ? index : null;
  }

  // ─────────────────────────────
  // Simple semver-ish "greater than" for version tie-breaking across TFMs.
  // Falls back to lexicographic comparison for non-semver strings.
  // ─────────────────────────────
  static _versionGt(a, b) {
    const parse = v => v.split(/[.\-]/).map(p => parseInt(p, 10) || 0);
    const pa = parse(a), pb = parse(b);
    for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
      const diff = (pa[i] ?? 0) - (pb[i] ?? 0);
      if (diff !== 0) return diff > 0;
    }
    return false;
  }

  // ─────────────────────────────
  // Parse obj/project.assets.json (fallback).
  // Returns Map<lowercaseName, { name, version, type, dependencies[] }> or null.
  // ─────────────────────────────
  static _readProjectAssets(projectRoot) {
    const assetPath = path.join(projectRoot, "obj", "project.assets.json");
    if (!fs.existsSync(assetPath)) return null;

    let raw;
    try {
      raw = JSON.parse(fs.readFileSync(assetPath, "utf8"));
    } catch {
      return null;
    }

    const index = new Map();

    for (const [libKey, meta] of Object.entries(raw.libraries ?? {})) {
      const slash   = libKey.lastIndexOf("/");
      const name    = slash >= 0 ? libKey.slice(0, slash)  : libKey;
      const version = slash >= 0 ? libKey.slice(slash + 1) : "";
      if (!version) continue;                     // skip malformed entries
      const key = name.toLowerCase();

      if (!index.has(key)) {
        index.set(key, {
          name,
          version,
          type:         (meta.type ?? "package").toLowerCase(),
          dependencies: Object.keys(meta.dependencies ?? {}).map(d => d.toLowerCase())
        });
      }
    }

    return index.size ? index : null;
  }

  // ─────────────────────────────
  // Read direct deps from .csproj.
  // Handles both single-line and multi-line PackageReference elements.
  // Returns { prod: Set<string>, dev: Set<string> } (lowercase names).
  // Falls back to Directory.Packages.props for version resolution when needed.
  // ─────────────────────────────
  static _readCsprojDeps(projectRoot) {
    const prod = new Set();
    const dev  = new Set();

    let projFile;
    try {
      projFile = fs.readdirSync(projectRoot).find(f => /\.(cs|fs|vb)proj$/.test(f));
    } catch {
      return { prod, dev };
    }
    if (!projFile) return { prod, dev };

    let xml;
    try {
      xml = fs.readFileSync(path.join(projectRoot, projFile), "utf8");
    } catch {
      return { prod, dev };
    }

    // Match both single-line and multi-line PackageReference elements.
    // [\s\S]*? spans newlines; stops at the first closing > or />
    const re = /<PackageReference\b([\s\S]*?)(?:\/>|>[\s\S]*?<\/PackageReference>)/gi;
    let m;
    while ((m = re.exec(xml)) !== null) {
      const attrs     = m[1];
      const nameMatch = attrs.match(/\bInclude=["']([^"']+)["']/i);
      if (!nameMatch) continue;

      const name      = nameMatch[1].toLowerCase();
      const condition = (attrs.match(/\bCondition=["']([^"']+)["']/i)?.[1] ?? "").toLowerCase();

      if (condition && (condition.includes("debug") || condition.includes("test"))) {
        dev.add(name);
      } else {
        prod.add(name);
      }
    }

    return { prod, dev };
  }

  // ─────────────────────────────
  // Scan a single .NET project directory.
  // ─────────────────────────────
  static _scanProject(projectRoot) {
    const index =
      this._readPackagesLock(projectRoot) ??
      this._readProjectAssets(projectRoot);

    if (!index) return [];

    const components = [];

    for (const [key, { name, version, type, dependencies }] of index.entries()) {
      const id = this._nugetPurl(name, version);

      const resolvedDeps = dependencies.map(dep => {
        const resolved = index.get(dep);
        return resolved
          ? this._nugetPurl(resolved.name, resolved.version)
          : this._nugetPurl(dep, "");
      });

      components.push({
        id,
        name:         key,
        version,
        type:         "library",
        license:      "unknown",
        ecosystem:    "csharp",
        state:        version ? "undetermined" : "version_unknown",
        scopes:       [],
        dependencies: resolvedDeps,
        paths:        [projectRoot],
        project_root: projectRoot,
        _nugetType:   type          // "direct" | "transitive" — used for scope seeding
      });
    }

    return components;
  }

  // ─────────────────────────────
  // Assign scopes (prod / dev) via BFS from direct deps declared in .csproj.
  //
  // Key fix: nameIdx is built PER PROJECT inside the loop, so BFS can never
  // escape into a different project's components.
  // ─────────────────────────────
  static _assignScopes(inventory) {
    const byId = new Map(inventory.map(c => [c.id, c]));

    for (const comp of inventory) {
      if (!Array.isArray(comp.scopes)) comp.scopes = [];
    }

    // Group components by project root
    const projectGroups = new Map();
    for (const comp of inventory) {
      const root = comp.project_root;
      if (!projectGroups.has(root)) projectGroups.set(root, []);
      projectGroups.get(root).push(comp);
    }

    for (const [projectRoot, comps] of projectGroups.entries()) {

      // Build a name index scoped ONLY to this project's components
      const nameIdx = new Map();
      for (const c of comps) {
        const existing = nameIdx.get(c.name) ?? [];
        existing.push(c);
        nameIdx.set(c.name, existing);
      }

      const { prod, dev } = this._readCsprojDeps(projectRoot);

      // Fallback: seed from _nugetType === "direct" only when the csproj file
      // could not be read (both sets empty AND no csproj found), not merely
      // when the csproj parsed to zero PackageReferences.
      if (prod.size === 0 && dev.size === 0) {
        let csprojExists = false;
        try {
          csprojExists = fs.readdirSync(projectRoot).some(f => /\.(cs|fs|vb)proj$/.test(f));
        } catch { /* ignore */ }

        if (!csprojExists) {
          for (const c of comps) {
            if (c._nugetType === "direct") prod.add(c.name);
          }
        }
      }

      const propagate = (names, scope) => {
        const queue = [];
        for (const n of names) {
          for (const c of (nameIdx.get(n) ?? [])) queue.push(c);
        }

        const visited = new Set();
        while (queue.length) {
          const c = queue.shift();
          if (visited.has(c.id)) continue;
          visited.add(c.id);

          if (!c.scopes.includes(scope)) c.scopes.push(scope);

          for (const depId of c.dependencies) {
            const d = byId.get(depId);
            // Only follow edges within this project
            if (d && d.project_root === projectRoot) queue.push(d);
          }
        }
      };

      propagate(prod, "prod");
      propagate(dev,  "dev");

      // Fallback — any component still unscoped defaults to prod
      for (const c of comps) {
        if (c.scopes.length === 0) c.scopes.push("prod");
      }
    }
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }
      const existing = map.get(comp.id);
      for (const p of comp.paths)   { if (!existing.paths.includes(p))   existing.paths.push(p); }
      for (const s of comp.scopes)  { if (!existing.scopes.includes(s))  existing.scopes.push(s); }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // Walks startDir (and subdirs) for .NET project roots, then scans each.
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {
    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    // Scan startDir itself if it is a .NET root
    if (CSharpNuGetScanner._isDotnetRoot(startDir)) {
      const key = path.resolve(startDir);
      visited.add(key);
      raw.push(...CSharpNuGetScanner._scanProject(startDir));
    }

    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (["node_modules", ".git", ".ubel", "obj", "bin", "packages"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);
        const key  = path.resolve(full);

        if (CSharpNuGetScanner._isDotnetRoot(full)) {
          if (!visited.has(key)) {
            visited.add(key);
            raw.push(...CSharpNuGetScanner._scanProject(full));
          }
          continue;
        }

        walk(full);
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this._assignScopes(merged);

    // Remove internal field before exposing inventory
    for (const c of merged) delete c._nugetType;

    this.inventoryData = merged;
    return merged.map(c => c.id);
  }
}

export default CSharpNuGetScanner;

File: python/ubel/csharp_runner.py
"""
csharp_scanner.py — C# / NuGet scanner (getInstalled only).

Mirrors the JS CSharpNuGetScanner:
  - Detects .NET project roots (*.csproj / *.fsproj / *.vbproj)
  - Reads packages.lock.json (preferred) or obj/project.assets.json
  - Assigns prod/dev scopes via BFS from .csproj PackageReference declarations
  - Merges duplicate PURLs across multi-project repos
  - PURL: pkg:nuget/<name>@<version>
"""

from __future__ import annotations

import json
import os
import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Set, Tuple


class CSharpNuGetScanner:

    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    def _nuget_purl(self,name: str, version: str) -> str:
        base = f"pkg:nuget/{name.lower()}"
        return f"{base}@{version}" if version else base

    # ------------------------------------------------------------------ #
    # Detect .NET root                                                     #
    # ------------------------------------------------------------------ #

    def _is_dotnet_root(directory: str) -> bool:
        try:
            return any(
                re.search(r"\.(cs|fs|vb)proj$", f)
                for f in os.listdir(directory)
            )
        except OSError:
            return False

    # ------------------------------------------------------------------ #
    # Central Package Management (Directory.Packages.props)               #
    # ------------------------------------------------------------------ #

    def _read_central_package_props(start_dir: str) -> Dict[str, str]:
        versions: Dict[str, str] = {}
        directory = start_dir
        for _ in range(8):
            candidate = os.path.join(directory, "Directory.Packages.props")
            if os.path.exists(candidate):
                try:
                    xml = Path(candidate).read_text(encoding="utf-8")
                except OSError:
                    break
                pattern = re.compile(
                    r'<PackageVersion[\s\S]*?Include=["\']([^"\']+)["\'][\s\S]*?Version=["\']([^"\']+)["\']',
                    re.IGNORECASE,
                )
                for m in pattern.finditer(xml):
                    versions[m.group(1).lower()] = m.group(2)
                break
            parent = os.path.dirname(directory)
            if parent == directory:
                break
            directory = parent
        return versions

    # ------------------------------------------------------------------ #
    # Version comparison                                                   #
    # ------------------------------------------------------------------ #

    def _version_gt(a: str, b: str) -> bool:
        def parse(v: str) -> List[int]:
            return [int(p) if p.isdigit() else 0 for p in re.split(r"[.\-]", v)]
        pa, pb = parse(a), parse(b)
        for i in range(max(len(pa), len(pb))):
            ai = pa[i] if i < len(pa) else 0
            bi = pb[i] if i < len(pb) else 0
            if ai != bi:
                return ai > bi
        return False

    # ------------------------------------------------------------------ #
    # packages.lock.json                                                   #
    # ------------------------------------------------------------------ #

    def _read_packages_lock(self,project_root: str) -> Optional[Dict[str, Any]]:
        lock_path = os.path.join(project_root, "packages.lock.json")
        if not os.path.exists(lock_path):
            return None
        try:
            raw = json.loads(Path(lock_path).read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return None

        index: Dict[str, Any] = {}
        for tfm_deps in raw.get("dependencies", {}).values():
            for pkg_id, meta in tfm_deps.items():
                key = pkg_id.lower()
                version = meta.get("resolved") or meta.get("requested") or ""
                if not version:
                    continue
                if key not in index:
                    index[key] = {
                        "name": pkg_id,
                        "version": version,
                        "type": (meta.get("type") or "direct").lower(),
                        "dependencies": [d.lower() for d in meta.get("dependencies", {})],
                    }
                else:
                    if self._version_gt(version, index[key]["version"]):
                        index[key]["version"] = version

        return index if index else None

    # ------------------------------------------------------------------ #
    # obj/project.assets.json                                              #
    # ------------------------------------------------------------------ #

    def _read_project_assets(project_root: str) -> Optional[Dict[str, Any]]:
        asset_path = os.path.join(project_root, "obj", "project.assets.json")
        if not os.path.exists(asset_path):
            return None
        try:
            raw = json.loads(Path(asset_path).read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return None

        index: Dict[str, Any] = {}
        for lib_key, meta in raw.get("libraries", {}).items():
            slash = lib_key.rfind("/")
            name    = lib_key[:slash]  if slash >= 0 else lib_key
            version = lib_key[slash+1:] if slash >= 0 else ""
            if not version:
                continue
            key = name.lower()
            if key not in index:
                index[key] = {
                    "name": name,
                    "version": version,
                    "type": (meta.get("type") or "package").lower(),
                    "dependencies": [d.lower() for d in meta.get("dependencies", {})],
                }

        return index if index else None

    # ------------------------------------------------------------------ #
    # .csproj direct deps                                                  #
    # ------------------------------------------------------------------ #

    def _read_csproj_deps(self,project_root: str) -> Tuple[Set[str], Set[str]]:
        prod: Set[str] = set()
        dev:  Set[str] = set()

        try:
            proj_file = next(
                (f for f in os.listdir(project_root) if re.search(r"\.(cs|fs|vb)proj$", f)),
                None,
            )
        except OSError:
            return prod, dev

        if not proj_file:
            return prod, dev

        try:
            xml = Path(os.path.join(project_root, proj_file)).read_text(encoding="utf-8")
        except OSError:
            return prod, dev

        pattern = re.compile(
            r"<PackageReference\b([\s\S]*?)(?:/>|>[\s\S]*?</PackageReference>)",
            re.IGNORECASE,
        )
        for m in pattern.finditer(xml):
            attrs = m.group(1)
            name_m = re.search(r'\bInclude=["\']([^"\']+)["\']', attrs, re.IGNORECASE)
            if not name_m:
                continue
            name = name_m.group(1).lower()
            cond_m = re.search(r'\bCondition=["\']([^"\']+)["\']', attrs, re.IGNORECASE)
            condition = (cond_m.group(1) if cond_m else "").lower()
            if condition and ("debug" in condition or "test" in condition):
                dev.add(name)
            else:
                prod.add(name)

        return prod, dev

    # ------------------------------------------------------------------ #
    # Scan one project                                                     #
    # ------------------------------------------------------------------ #

    def _scan_project(self,project_root: str) -> List[Dict[str, Any]]:
        index = (
            self._read_packages_lock(project_root)
            or self._read_project_assets(project_root)
        )
        if not index:
            return []

        components: List[Dict[str, Any]] = []

        for key, entry in index.items():
            name    = entry["name"]
            version = entry["version"]
            cid     = self._nuget_purl(name, version)

            resolved_deps: List[str] = []
            for dep in entry["dependencies"]:
                if dep in index:
                    d = index[dep]
                    resolved_deps.append(self._nuget_purl(d["name"], d["version"]))
                else:
                    resolved_deps.append(self._nuget_purl(dep, ""))

            components.append({
                "id":           cid,
                "name":         key,
                "version":      version,
                "type":         "library",
                "license":      "unknown",
                "ecosystem":    "csharp",
                "state":        "undetermined" if version else "version_unknown",
                "scopes":       [],
                "dependencies": resolved_deps,
                "paths":        [project_root],
                "project_root": project_root,
                "_nuget_type":  entry["type"],
            })

        return components

    # ------------------------------------------------------------------ #
    # BFS scope propagation                                                #
    # ------------------------------------------------------------------ #

    def _assign_scopes(self,inventory: List[Dict[str, Any]]) -> None:
        by_id: Dict[str, Dict] = {c["id"]: c for c in inventory}

        for comp in inventory:
            if not isinstance(comp.get("scopes"), list):
                comp["scopes"] = []

        # Group by project root
        project_groups: Dict[str, List[Dict]] = {}
        for comp in inventory:
            root = comp["project_root"]
            project_groups.setdefault(root, []).append(comp)

        for project_root, comps in project_groups.items():
            name_idx: Dict[str, List[Dict]] = {}
            for c in comps:
                name_idx.setdefault(c["name"], []).append(c)

            prod, dev = self._read_csproj_deps(project_root)

            if not prod and not dev:
                csproj_exists = False
                try:
                    csproj_exists = any(
                        re.search(r"\.(cs|fs|vb)proj$", f)
                        for f in os.listdir(project_root)
                    )
                except OSError:
                    pass
                if not csproj_exists:
                    for c in comps:
                        if c.get("_nuget_type") == "direct":
                            prod.add(c["name"])

            def propagate(names: Set[str], scope: str) -> None:
                queue = [c for n in names for c in name_idx.get(n, [])]
                visited: Set[str] = set()
                while queue:
                    c = queue.pop(0)
                    if c["id"] in visited:
                        continue
                    visited.add(c["id"])
                    if scope not in c["scopes"]:
                        c["scopes"].append(scope)
                    for dep_id in c["dependencies"]:
                        d = by_id.get(dep_id)
                        if d and d.get("project_root") == project_root:
                            queue.append(d)

            propagate(prod, "prod")
            propagate(dev,  "dev")

            for c in comps:
                if not c["scopes"]:
                    c["scopes"].append("prod")

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    def merge_inventory_by_purl(self,components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        if self._is_dotnet_root(start_dir):
            visited.add(start_dir)
            raw.extend(self._scan_project(start_dir))

        skip = {"node_modules", ".git", ".ubel", "obj", "bin", "packages"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    key  = os.path.realpath(full)
                    if self._is_dotnet_root(full):
                        if key not in visited:
                            visited.add(key)
                            raw.extend(self._scan_project(full))
                        continue
                    walk(full)

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self._assign_scopes(merged)

        for c in merged:
            c.pop("_nuget_type", None)

        self.inventory_data = merged
        return [c["id"] for c in merged]